rust/list/src/lib.rs

```diff
@@ -346,10 +370,73 @@ pub unsafe extern "C" fn list_manager_get_all_labels(manager: *const ListManager
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn list_manager_create_item(manager: *mut ListManager, item: *const Item) {
+pub unsafe extern "C" fn list_manager_create_item(manager: *mut ListManager, name: *const c_char, due_date: *const time_t) {
+    let name = c_char_to_string(name);
+    log::d(&format!("Creating item: {:?}, {:?}, {:?}", name, due_date, manager)[..]);
+
     let manager = &mut*manager;
-    let item = &*item;
-    let _ = manager.create_item(&item);
+    let mut item = Item::default();
+
+    item.name = name;
+    item.due_date = Some(Timespec::new(due_date as i64, 0));
+
+    manager.create_item(&item);
+ 
+    // Let whoever's listening know that there are new items!
+    match CHANGED_CALLBACK {
+        Some(f) => f(),
+        None => ()
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct ItemCSet {
+    items: Box<[ItemC]>
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn list_manager_on_items_changed(callback: extern fn()) {
+    CHANGED_CALLBACK = Some(callback);
+    callback();
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn list_manager_all_items(manager: *mut ListManager, callback: extern "C" fn(Option<&ItemCSet>)) {
+    let manager = &*manager;
+    let items: ItemsC = manager.fetch_items().map(|item| item.into()).expect("all items");
+
+    // TODO there's bound to be a better way. Ideally this should just return an empty set,
+    // but I ran into problems while doing that.
+    let count = items.vec.len();
+
+    let set = ItemCSet {
+        items: items.vec.into_boxed_slice()
+    };
+
+    let res = match count > 0 {
+        // NB: we're lending a set, it will be cleaned up automatically once 'callback' returns
+        true => Some(&set),
+        false => None
+    };
+
+    callback(res);
+
+    // TODO can we just clean up after ourselves here, instead of in item_c_destroy?
```

Comment:
Not advisable. If we clean up memory here while our client still has ownership of the information then we could end up with some problems when we come to iterate our items.